### PR TITLE
feat: add wasm light client module to feat/wasm-clients branch

### PR DESCRIPTION
**Pull request by misko9**
## Description

Wasm light client module

This PR adds a new light client to support wasm light client contracts. A chain that would like to use a wasm light client contract would add this module to their app.go. An existing chain will need to add 08-wasm to the allowed clients list. The module will create a wasm vm to run the wasm contracts. New contracts will be submitted through a governance proposal. Once approved, the contract can be used to create any number of clients. Gas consumption for the wasm vm is based off of wasmd's implementation. The bulk of the light client logic is executed in the contract with the module providing the interface and management of the contracts.

IBC spec: https://github.com/cosmos/ibc/tree/main/spec/client/ics-008-wasm-client

### Commit Message / Changelog Entry

```bash
feat: adding 08-wasm light client module
```

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package). Unit tests have been written providing code coverage of the interface to a contract and the management of contracts. Currently, a version of the grandpa light client contract is being used with a couple methods stubbed out as they are still in development, but the interface is still fully exercised.
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/02-client/keeper/keeper.go [hunk 1]
```diff
@@ -270,18 +271,38 @@ func (k Keeper) GetSelfConsensusState(ctx sdk.Context, height exported.Height) (
...
+	}
+
 	return consensusState, nil
 }
 
 // ValidateSelfClient validates the client parameters for a client of the running chain
 // This function is only used to validate the client state the counterparty stores for this chain
 // Client must be in same revision as the executing chain
 func (k Keeper) ValidateSelfClient(ctx sdk.Context, clientState exported.ClientState) error {
+	if clientState.ClientType() == exported.Wasm {
```
[the remark this answers is not in the document]
Comment (misko9) [reply modules/core/02-client/keeper/keeper.go:299]: If the client type is `08-wasm`, then the tendermint client state is already in `data`. We are unmarshaling the tendermint client state from the wasm client state in order to validate it.

[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/core/02-client/keeper/keeper.go:299]: Won't any other non-tendermint wasm client states just fail here on line 306, which would bubble up the error to connOpenTry?
Am I missing something?

[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/core/02-client/keeper/keeper.go:299]: Ayy, indeed! Makes sense. We'll definitely need future improvements to `ValidateSelfClient` in a world where sdk supports various consensus engines.

**Review by crodriguezvega — APPROVED** [review]
Thank so much for the fantastic work implementing this new light client module, @misko9. 🙏 Thank you also, @damiannolan and @DimitrisJim for the review of the PR and the rest of the team for spending hours also reviewing the code during our internal walkthrough. I will merge now this PR and we will continue with the remaining work on follow-up PRs.

**Pull request merged**